src/test/java/org/cactoos/iterator/SyncIteratorTest.java

```diff
@@ -73,4 +78,201 @@ public void syncIteratorReturnsCorrectValuesWithInternalLock() {
+            final Runnable second = () -> {
+                sync.add(iterator.next());
+            };
+            new Concurrent(first, second).launch();
+            MatcherAssert.assertThat(
+                "Missing the list items(s) (next()).",
+                sync,
+                Matchers.containsInAnyOrder("a", "b")
+            );
+        }
+    }
+
+    @Test
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public void correctValuesForConcurrentNextHasNext()
+        throws InterruptedException {
+        for (int iter = 0; iter < 5000; iter += 1) {
+            final List<String> list = Arrays.asList("a", "b");
+            final SyncIterator<String> iterator = new SyncIterator<>(
+                list.iterator()
+            );
+            final List<Object> sync =
+                Collections.synchronizedList(
+                    new ArrayList<>(list.size())
+                );
+            final Runnable first = () -> {
+                sync.add(iterator.next());
+            };
+            final Runnable second = () -> {
+                sync.add(iterator.next());
+            };
+            final Runnable third = () -> {
+                sync.add(iterator.hasNext());
+            };
+            new Concurrent(first, second, third).launch();
+            MatcherAssert.assertThat(
+                "Missing the list items(s) (next()).",
+                sync,
+                Matchers.allOf(
+                    Matchers.hasItem("a"),
+                    Matchers.hasItem("b")
+                )
+            );
+            MatcherAssert.assertThat(
+                "Missing hasNext() value.",
+                sync,
+                Matchers.anyOf(
+                    Matchers.hasItem(true),
+                    Matchers.hasItem(false)
+                )
+            );
+        }
+    }
+
+    @Test
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public void correctValuesForConcurrentHasNextHasNext()
+        throws InterruptedException {
+        for (int iter = 0; iter < 5000; iter += 1) {
+            final List<String> list = Arrays.asList("a", "b");
+            final SyncIterator<String> iterator = new SyncIterator<>(
+                list.iterator()
+            );
+            final List<Object> sync =
+                Collections.synchronizedList(
+                    new ArrayList<>(list.size())
+                );
+            final Runnable first = () -> {
+                sync.add(iterator.hasNext());
+            };
+            final Runnable second = () -> {
+                sync.add(iterator.hasNext());
+            };
+            new Concurrent(first, second).launch();
+            MatcherAssert.assertThat(
+                "Missing hasNext() value(s).",
+                sync,
+                Matchers.contains(true, true)
+            );
+        }
+    }
+
+    /**
+     * Tests runnables for concurrency issues.
+     */
+    private final class Concurrent {
```

Comment:
@driver733 I have a big suspicion that you are reinventing here what we already have in `RunInThreads`. Can you please check?